django/tasks/backends/base.py

```diff
@@ -0,0 +1,138 @@
+from abc import ABCMeta, abstractmethod
+from inspect import iscoroutinefunction
+
+from asgiref.sync import sync_to_async
+
+from django.conf import settings
+from django.core import checks
+from django.db import connections
+from django.tasks import DEFAULT_TASK_QUEUE_NAME
+from django.tasks.base import (
+    DEFAULT_TASK_PRIORITY,
+    TASK_MAX_PRIORITY,
+    TASK_MIN_PRIORITY,
+    Task,
+)
+from django.tasks.exceptions import InvalidTask
+from django.utils import timezone
+from django.utils.inspect import get_func_args, is_module_level_function
+
+
+class BaseTaskBackend(metaclass=ABCMeta):
+    task_class = Task
+
+    # Does the backend support Tasks to be enqueued with the run_after
+    # attribute?
+    supports_defer = False
+
+    # Does the backend support coroutines to be enqueued?
+    supports_async_task = False
+
+    # Does the backend support results being retrieved (from any
+    # thread/process)?
+    supports_get_result = False
+
+    # Does the backend support executing Tasks in a given
+    # priority order?
+    supports_priority = False
+
+    def __init__(self, alias, params):
+        self.alias = alias
+        self.queues = set(params.get("QUEUES", [DEFAULT_TASK_QUEUE_NAME]))
+        self.enqueue_on_commit = bool(params.get("ENQUEUE_ON_COMMIT", True))
+        self.options = params.get("OPTIONS", {})
+
+    def _get_enqueue_on_commit_for_task(self, task):
+        return (
+            task.enqueue_on_commit
+            if task.enqueue_on_commit is not None
+            else self.enqueue_on_commit
+        )
```

Comment:
Late to the party here but the notion of enqueue on commit is broken in its current form as it's not multi-database aware. In other words, it's okay for this feature to default to the default database but there should be a way to tie it to a particular `DATABASES` entry (or entries) as not all writes that pertain to task scheduling will necessarily be tied to the default database.

Reply:
This will become even apparent on a theorical database backed task backend which might use a dedicated `DATABASES` alias for storage purposes as it means the task row might now be persisted when the worker attempts to picks it up.

I understand the desire to automatically euqueue tasks when appropriate but I think we might need to get back to the drawing board for this particular aspect and encourage usage of `transaction.on_commit(partial(tasks.enqueue, ...), alias)` instead of providing `ENQUEUE_ON_COMMIT` instead?